From 8f4e7848ea5587633d3703309e6dafba27f02820 Mon Sep 17 00:00:00 2001
From: Ulf Hermann <ulf.hermann@qt.io>
Date: Fri, 10 Oct 2025 10:47:46 +0200
Subject: [PATCH 146/239] QtQml: Fix resolution of aliases to local aliases

When we find an alias to a local alias, we need to first check if the
target alias is already resolved. Otherwise we can't insert the source
alias, yet.

Amends commit 9e1378260a3f3a6eb4cad8d6ae47bc218e6a2b5f.

Fixes: QTBUG-141059
Change-Id: If272a746595b39cb2dbc2b8cc0c2f688277cfe59
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit 75dcbb156b2b1dfe0bb43e3119735f26f848e0f6)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qml/qml/qqmltypecompiler.cpp              | 25 +++++++++++++++----
 .../qml/qmlcppcodegen/data/CMakeLists.txt     |  1 +
 .../qmlcppcodegen/data/aliasToLocalAlias.qml  | 11 ++++++++
 .../qml/qmlcppcodegen/tst_qmlcppcodegen.cpp   | 22 ++++++++++++++++
 .../qqmllanguage/data/cyclicAlias.errors.txt  |  2 +-
 5 files changed, 55 insertions(+), 6 deletions(-)
 create mode 100644 tests/auto/qml/qmlcppcodegen/data/aliasToLocalAlias.qml

diff --git a/src/qml/qml/qqmltypecompiler.cpp b/src/qml/qml/qqmltypecompiler.cpp
index 509deed996..ff2cc87ff1 100644
--- a/src/qml/qml/qqmltypecompiler.cpp
+++ b/src/qml/qml/qqmltypecompiler.cpp
@@ -847,6 +847,7 @@ QQmlComponentAndAliasResolver<QQmlTypeCompiler>::resolveAliasesInObject(
 
     int aliasIndex = 0;
     int numSkippedAliases = 0;
+    bool hasUnresolvedLocalAliases = false;
 
     for (QmlIR::Alias *alias = obj->firstAlias(); alias; alias = alias->next, ++aliasIndex) {
         if (resolvedAliases.contains(alias)) {
@@ -903,14 +904,25 @@ QQmlComponentAndAliasResolver<QQmlTypeCompiler>::resolveAliasesInObject(
             if (!targetProperty) {
                 bool aliasPointsToOtherAlias = false;
                 int localAliasIndex = 0;
-                for (auto targetAlias = targetObject->aliasesBegin(), end = targetObject->aliasesEnd(); targetAlias != end; ++targetAlias, ++localAliasIndex) {
+                auto targetAlias = targetObject->aliasesBegin();
+                for (const auto end = targetObject->aliasesEnd(); targetAlias != end;
+                        ++targetAlias, ++localAliasIndex) {
                     if (stringAt(targetAlias->nameIndex()) == property) {
                         aliasPointsToOtherAlias = true;
                         break;
                     }
                 }
                 if (aliasPointsToOtherAlias) {
-                    if (targetObjectIndex == objectIndex) {
+                    if (targetObjectIndex != objectIndex) {
+                        // Don't continue, yet. We need to respect the order of objects.
+                        alias->setIdIndex(idIndex);
+                        return aliasIndex == numSkippedAliases
+                                ? NoAliasResolved
+                                : SomeAliasesResolved;
+                    }
+
+                    if (resolvedAliases.contains(targetAlias)) {
+                        // Target already resolved. We can set the alias right away.
                         alias->localAliasIndex = localAliasIndex;
                         alias->setIsAliasToLocalAlias(true);
                         if (!appendAliasToPropertyCache(
@@ -921,10 +933,13 @@ QQmlComponentAndAliasResolver<QQmlTypeCompiler>::resolveAliasesInObject(
                         continue;
                     }
 
-                    // restore
+                    // Target isn't resolved yet, but it's in the same object.
+                    // Continue with the other aliases.
                     alias->setIdIndex(idIndex);
                     // Try again later and resolve the target alias first.
-                    return aliasIndex == numSkippedAliases ? NoAliasResolved : SomeAliasesResolved;
+                    ++numSkippedAliases;
+                    hasUnresolvedLocalAliases = true;
+                    continue;
                 }
             }
 
@@ -992,7 +1007,7 @@ QQmlComponentAndAliasResolver<QQmlTypeCompiler>::resolveAliasesInObject(
     if (numSkippedAliases == aliasIndex)
         return NoAliasResolved;
 
-    if (aliasIndex == obj->aliasCount())
+    if (aliasIndex == obj->aliasCount() && !hasUnresolvedLocalAliases)
         return AllAliasesResolved;
 
     return SomeAliasesResolved;
diff --git a/tests/auto/qml/qmlcppcodegen/data/CMakeLists.txt b/tests/auto/qml/qmlcppcodegen/data/CMakeLists.txt
index c30e70cc22..b59472c9c5 100644
--- a/tests/auto/qml/qmlcppcodegen/data/CMakeLists.txt
+++ b/tests/auto/qml/qmlcppcodegen/data/CMakeLists.txt
@@ -98,6 +98,7 @@ set(qml_files
     WindowDerived.qml
     aliasLookup.qml
     aliasToAliasResolutionSkipCorruption.qml
+    aliasToLocalAlias.qml
     ambiguous1/Ambiguous.qml
     ambiguous2/Ambiguous.qml
     ambiguousAs.qml
diff --git a/tests/auto/qml/qmlcppcodegen/data/aliasToLocalAlias.qml b/tests/auto/qml/qmlcppcodegen/data/aliasToLocalAlias.qml
new file mode 100644
index 0000000000..6bd4b20679
--- /dev/null
+++ b/tests/auto/qml/qmlcppcodegen/data/aliasToLocalAlias.qml
@@ -0,0 +1,11 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR GPL-3.0-only
+pragma Strict
+import QtQml
+
+QtObject {
+    id: option
+
+    property alias label: option.description
+    property alias description: option.objectName
+}
diff --git a/tests/auto/qml/qmlcppcodegen/tst_qmlcppcodegen.cpp b/tests/auto/qml/qmlcppcodegen/tst_qmlcppcodegen.cpp
index ba1d108112..cfca8206f6 100644
--- a/tests/auto/qml/qmlcppcodegen/tst_qmlcppcodegen.cpp
+++ b/tests/auto/qml/qmlcppcodegen/tst_qmlcppcodegen.cpp
@@ -48,6 +48,7 @@ private slots:
     void accessModelMethodFromOutSide();
     void aliasLookup();
     void aliasToAliasResolutionSkipCorruption();
+    void aliasToLocalAlias();
     void ambiguousAs();
     void ambiguousSignals();
     void anchorsFill();
@@ -610,6 +611,27 @@ void tst_QmlCppCodegen::aliasToAliasResolutionSkipCorruption()
     QVERIFY(!object.isNull());
 }
 
+void tst_QmlCppCodegen::aliasToLocalAlias()
+{
+    QQmlEngine engine;
+    QQmlComponent component(&engine, QUrl(u"qrc:/qt/qml/TestTypes/aliasToLocalAlias.qml"_s));
+    QVERIFY2(!component.isError(), component.errorString().toUtf8());
+    QScopedPointer<QObject> object(component.create());
+    QVERIFY(!object.isNull());
+
+    const QVariant label = object->property("label");
+    QCOMPARE(label.metaType(), QMetaType::fromType<QString>());
+    QVERIFY(label.toString().isEmpty());
+
+    const QVariant description = object->property("description");
+    QCOMPARE(description.metaType(), QMetaType::fromType<QString>());
+    QVERIFY(description.toString().isEmpty());
+
+    object->setObjectName(u"nnn");
+    QCOMPARE(object->property("label").toString(), u"nnn");
+    QCOMPARE(object->property("description").toString(), u"nnn");
+}
+
 void tst_QmlCppCodegen::ambiguousAs()
 {
     QQmlEngine e;
diff --git a/tests/auto/qml/qqmllanguage/data/cyclicAlias.errors.txt b/tests/auto/qml/qqmllanguage/data/cyclicAlias.errors.txt
index 46951ef69f..812ca78b03 100644
--- a/tests/auto/qml/qqmllanguage/data/cyclicAlias.errors.txt
+++ b/tests/auto/qml/qqmllanguage/data/cyclicAlias.errors.txt
@@ -1 +1 @@
-4:5:Cyclic alias
+4:5:Circular alias reference detected
-- 
2.51.2

